### PR TITLE
Added experimental disable_reuseaddr to bypass the issue

### LINKED ISSUE #2819: mirrord stealing to different app
Some of our users want to use mirrord over `kubectl proxy` which runs on a bastion host they SSH and port-forward to.
They're running 
`ssh -L 8080:localhost:8080`
then inside the ssh
`kubectl proxy`

then on their local machine
```
kubectl config set-cluster testt --server=http://localhost:8080 --insecure-skip-tls-verify=true
kubectl config set-context testt --cluster=testt
kubectl config use-context testt
```

then they `mirrord exec` a Java Bootspring app that listens on 8080, requests gets stolen but when it's stolen it is sent to... the kubectl proxy (ssh forward) instead of the app :|
happens on macOS

ssh version:
OpenSSH_9.7p1, LibreSSL 3.3.6

lsof output (`lsof -nP -iTCP:8080 -sTCP:LISTEN`):
```
ssh     36023 user    7u  IPv6 0x.ccc      0t0  TCP [::1]:8080 (LISTEN)

ssh     36023 user    8u  IPv4  0xbb...      0t0  TCP [127.0.0.1:8080](http://127.0.0.1:8080/) (LISTEN)

nc      36503 user    5u  IPv4 0xaaa..      0t0  TCP *:8080 (LISTEN)
```

### PR DESCRIPTION
**Pull request by aviramha**
Closes #2819

### CODE REVIEW

File: mirrord/layer/src/socket/ops.rs [hunk 1]
```diff
@@ -275,6 +274,22 @@ pub(super) fn bind(
         Err(HookError::AddressAlreadyBound(requested_address))?;
     }
 
+    #[cfg(target_os = "macos")]
+    unsafe {
```
Comment (Razz4780) [inline mirrord/layer/src/socket/ops.rs:278]: Can we restrict the scope of this `unsafe` to bare minimum?

File: mirrord/layer/src/socket/ops.rs [hunk 2]
```diff
@@ -275,6 +274,22 @@ pub(super) fn bind(
...
+        let experimental = crate::setup().experimental();
+        if experimental.disable_reuseaddr {
+            tracing::debug!("here!!");
```
Comment (Razz4780) [inline mirrord/layer/src/socket/ops.rs:281]: Delet :wastebasket:

File: mirrord/layer/src/socket/ops.rs [hunk 3]
```diff
@@ -275,6 +274,22 @@ pub(super) fn bind(
...
+            let socket = socket2::Socket::from_raw_fd(sockfd);
+            if let Err(e) = socket.set_reuse_address(false) {
+                tracing::debug!(?e, "Failed to set SO_REUSEADDR to false");
+            }
+            if let Err(e) = socket.set_reuse_port(false) {
+                tracing::debug!(?e, "Failed to set SE_REUSEPORT to false");
+            }
+            // get ownership of fd back
+            let _ = socket.into_raw_fd();
```
Comment (Razz4780) [inline mirrord/layer/src/socket/ops.rs:290]: I'd use [nix function](https://docs.rs/nix/latest/nix/sys/socket/fn.setsockopt.html) and `BorrowedFd::borrow_raw` for this. Just a suggestion tho

**Review by Razz4780 — CHANGES_REQUESTED** [review]

File: mirrord/layer/src/socket/ops.rs (hunk 2 above)
Comment (gememma) [inline mirrord/layer/src/socket/ops.rs:281]: 🚨 👀 debug statement

File: mirrord-schema.json [hunk 4]
```diff
@@ -756,6 +756,14 @@
       "description": "mirrord Experimental features. This shouldn't be used unless someone from MetalBear/mirrord tells you to.",
       "type": "object",
       "properties": {
+        "disable_reuseaddr": {
+          "title": "_experimental_ disable_reuseaddr {#experimental-disable_reuseaddr}",
+          "description": "Disables the `SO_REUSEADDR` socket option on sockets that mirrord steals/mirrors. on macOS the application can use the same address many times but then we don't steal it correctly. This probably should be on by default but we want to gradual roll it out. https://github.com/metalbear-co/mirrord/issues/2819 This option applies only on macOS.",
```
Comment (gememma) [inline mirrord-schema.json:761]:
```suggestion
          "description": "Disables the `SO_REUSEADDR` socket option on sockets that mirrord steals/mirrors. On macOS the application can use the same address many times but then we don't steal it correctly. This probably should be on by default but we want to gradually roll it out. https://github.com/metalbear-co/mirrord/issues/2819 This option applies only on macOS.",
```

Comment (aviramha) [reply mirrord-schema.json:761]: Thanks for the suggestion! P.S, next time send the review on the comments where it's being generated from :D

File: mirrord/layer/src/socket/ops.rs (hunk 3 above)
Comment (aviramha) [reply mirrord/layer/src/socket/ops.rs]: nice!

**Review by Razz4780 — APPROVED** [review]

**Pull request merged**